python/phase-1.md

```diff
@@ -0,0 +1,441 @@
+* Be reasonably flexible with reasonable defaults
+* Support only Python >= 3.7
+* Be friendly to REPL/interpreter use cases
+
+The stack/approach looks like:
+
+* Python >= 3.7 (so may need `typing_extension` for things like `ParamSpec`)
+* Poetry packaging, black formatting
+  * Pipenv originally considered
+* PyO3 + sdk-core
+* asyncio direct (i.e. no need for anyio-style wrapper)
+* Type annotations everywhere
+* Decorator-based activity and workflow definitions
+
+An example of a client (intentionally missing _a lot_ of features):
+
+```python
+import temporalio
+
+
+async def main():
+    core = await temporalio.Core.default_local()
+    client = temporalio.Client(core)
+    result = await client.execute_workflow(
+        "MyWorkflow", "arg1", id="someid", task_queue="somequeue"
+    )
+    print("Result", result)
+```
+
+An example of an activity and registration (also missing a lot of features):
+
+```python
+import temporalio
+
+
+def say_hello(name: str) -> str:
+    return f"Hello, {name}!"
+
+
+async def main():
+    core = await temporalio.Core.default_local()
+    worker = temporalio.Worker(
+        core, task_queue="somequeue", activities={"say_hello": say_hello}
+    )
+    await worker.run()
+```
+
+## API Definitions
+
+A lot of loosely-defined API definitions are below. They are intentionally lacking details.
+
+High-level notes:
+
+* For demonstration/simplicity here we are leaving off a lot of proper typings (e.g. `Optional`) that the real version
+  will have
+* We leave off a lot of options on calls for this high-level proposal
+* We are not using proper function documentation, but rather inline comments for easier discussion
+* **NOTE: These will have proper docs and types when developed!**
+
+### Core
+
+Exported from `temporalio` (but probably in `temporalio/core/core.py`, exported from `temporalio.core` then
+exported from `temporalio`):
+
+```python
+class Core:
+    @staticmethod
+    async def create(
+        addr: str,
+        *,
+        namespace: str = "default"
+        # ... additional options omitted for brevity
+    ) -> Core:
+        pass
+
+    @classmethod
+    async def default_local(cls) -> Core:
+        pass
+
+    async def shutdown(self) -> None:
+        pass
+```
+
+Notes:
+
+* After discussion, it has been decided that both clients and workers will be backed by core
+    * Yes that means clients require the Rust library which we accept
+* After discussion, it has been decided that we will not have an implicit local default like TypeScript SDK, but instead
+  an explicit default
+  * `default_local()` memoizes success
+* Should we have a constructor, an `__aenter__`, and an `__aexit__`?
+  * Probably
+
+### Client
+
+Exported from `temporalio` (but probably in `temporalio/client/client.py`, exported from `temporalio.client` then
+exported from `temporalio`):
+
+```python
+class Client:
+    def __init__(core_or_service: temporalio.Core) -> None:
+        pass
```

Comment:
I don't think we want to init clients this way - I think `Core` should have a `get_client` method, and I think a client constructor should accept some client options object (which Core will also accept when created) which it will use to create a core-independent client via the publicly exposed interface here: https://github.com/temporalio/sdk-core/blob/e3f54276f971bcc9b6e1d1cf4d0ecbeb853387ae/client/src/lib.rs#L210

IE: You should be able to have a client tied to core if you want, but you should also be able to get a client without any core at all.

Obviously the rub is telemetry like we talked about a little bit. We can clean up how that works in core to make it easy, but we can start without any changes by simply having clients-not-from-core not do any telemetry.

Reply:
I had originally had similar, but:

* I didn't want to make an implementation of this class with non-core because I couldn't figure out how to return `service` properly as I don't think the core and non-core will have the same signatures for the call. So I am not sure I can have a core-independent client for all gRPC calls unless I maintain all gRPC calls manually. But I haven't dug enough to confirm.
* We don't have `Core.create_worker`, so I figured it was more consistent to pass core to clients/workers than to either be inconsistent or force worker creation to be on core too.

I personally think it's cleaner this way (and any other potential uses of the same core instance could potentially take core in their constructor too). But I can definitely change this if consensus is to do so.

Not too worried about the telemetry stuff, we can figure that out a bit later (I have some ideas that basically moves `tokio::tracing` to lang-side and requires core `init` to accept a spawner unfortunately since I can't see how to do thread-local on tokio runtime threads, but that's for a different setting).

Reply:
My main issue with tying Core to the Client is that they're 2 completely different entities and it's actually Core that takes a client and not the other way around on the rust side).

Reply:
I suppose it works (to use only a client in your python process) as long as you never register any workers. It's not really how I've thought of what Core is, but I guess it technically works. It's not super clean. You still have this core instance that you can "shut down" but that doesn't really mean anything, and it's got facilities for things you're never going to use if all you want is a client.

And "I couldn't figure out how to return service properly as I don't think the core and non-core will have the same signatures for the call. " is easily solved. It's not exposed right now in core but it's trivial to add, so I wouldn't let that stop you.

Regarding "Core.create_worker" -- why not? I mean I don't care a whole lot either way here but I think that makes sense too. If we want it to accept core that's fine still, and client can either accept core (`Client.from_core`), or can accept options for a core-independent client.

Reply:
> My main issue with tying Core to the Client is that they're 2 completely different entities and it's actually Core that takes a client and not the other way around on the rust side).

Can you clarify this? Are you talking about the off-PR discussion of whether to tie the client to core? I had understood that as settled.

> And "I couldn't figure out how to return service properly as I don't think the core and non-core will have the same signatures for the call. " is easily solved. It's not exposed right now in the rust client crate but it's trivial to add, so I wouldn't let that stop you.

I am not sure it's easily solved. If the Python gRPC client, say, takes contextual call options I can't expect to abstract those properly.

> Regarding "Core.create_worker" -- why not? I mean I don't care a whole lot either way here but I think that makes sense too. If we want it to accept core that's fine still, and client can either accept core (Client.from_core), or can accept options for a core-independent client.

I think it's cleaner to create instances of classes via Python's class creation mechanisms whenever reasonable.

Can I get clarity on what's being asked for here? So we want `get_client` and `create_worker` on core? Or we don't want tying Core to a client? Or we want an abstract client that works on pure Python gRPC and Core client including abstracting all workflowservice calls over both? I am looking for concrete complete suggestions (the piece here, even down to only have a Core version and to have the client and worker accept core in constructor non-defaulted was what was settled on during discussion, but maybe it was unclear and of course we can change).

Reply:
I am marking this resolved as we are having off-PR discussions about how core, clients, and workers interact.